selenium transient error error: [Errno 111] Connection refused

# Bug

## Description

functional tests fail to connect to firefox

## Steps to Reproduce

* pytest function in a loop on a very slow machine

## Expected Behavior

* tests pass

## Actual Behavior

* tests fail with a stack that looks like

<pre>
==================================== ERRORS ====================================
 ERROR at setup of TestJournalistLayout.test_account_new_two_factor_hotp[en_US] 
self = <test_journalist.TestJournalistLayout instance at 0x7f3964ee1560>
request = <SubRequest 'webdriver_fixture' for <Function 'test_account_new_two_factor_hotp[en_US]'>>
    @pytest.fixture(autouse=True, params=list_locales())
    def webdriver_fixture(self, request):
        self.accept_languages = request.param
        self.log_dir = abspath(
            os.path.join(dirname(realpath(__file__)),
                         'screenshots', self.accept_languages))
        os.system("mkdir -p " + self.log_dir)
        log_file = open(os.path.join(self.log_dir, 'firefox.log'), 'a')
        log_file.write(
            '\n\n[%s] Running Functional Tests\n' % str(
                datetime.now()))
        log_file.flush()
        firefox = firefox_binary.FirefoxBinary(log_file=log_file)
        profile = webdriver.FirefoxProfile()
        profile.set_preference("intl.accept_languages", self.accept_languages)
        self.override_driver = True
        self.driver = webdriver.Firefox(firefox_binary=firefox,
>                                       firefox_profile=profile)
tests/pages-layout/functional_test.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/firefox/webdriver.py:85: in __init__
    keep_alive=True)
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py:90: in __init__
    self.start_session(desired_capabilities, browser_profile)
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py:177: in start_session
    response = self.execute(Command.NEW_SESSION, capabilities)
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py:234: in execute
    response = self.command_executor.execute(driver_command, params)
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/remote_connection.py:401: in execute
    return self._request(command_info[0], url, body=data)
/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/remote_connection.py:432: in _request
    self._conn.request(method, parsed_url.path, body, headers)
/usr/lib/python2.7/httplib.py:1017: in request
    self._send_request(method, url, body, headers)
/usr/lib/python2.7/httplib.py:1051: in _send_request
    self.endheaders(body)
/usr/lib/python2.7/httplib.py:1013: in endheaders
    self._send_output(message_body)
/usr/lib/python2.7/httplib.py:864: in _send_output
    self.send(msg)
/usr/lib/python2.7/httplib.py:826: in send
    self.connect()
/usr/lib/python2.7/httplib.py:807: in connect
    self.timeout, self.source_address)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
address = ('127.0.0.1', '53440'), timeout = <object object at 0x7f3973662100>
source_address = None
    def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
                          source_address=None):
        """Connect to *address* and return the socket object.
    
        Convenience function.  Connect to *address* (a 2-tuple ``(host,
        port)``) and return the socket object.  Passing the optional
        *timeout* parameter will set the timeout on the socket instance
        before attempting to connect.  If no *timeout* is supplied, the
        global default timeout setting returned by :func:`getdefaulttimeout`
        is used.  If *source_address* is set it must be a tuple of (host, port)
        for the socket to bind as a source address before making the connection.
        An host of '' or port 0 tells the OS to use the default.
        """
    
        host, port = address
        err = None
        for res in getaddrinfo(host, port, 0, SOCK_STREAM):
            af, socktype, proto, canonname, sa = res
            sock = None
            try:
                sock = socket(af, socktype, proto)
                if timeout is not _GLOBAL_DEFAULT_TIMEOUT:
                    sock.settimeout(timeout)
                if source_address:
                    sock.bind(source_address)
                sock.connect(sa)
                return sock
    
            except error as _:
                err = _
                if sock is not None:
                    sock.close()
    
        if err is not None:
          raise err
E           error: [Errno 111] Connection refused
/usr/lib/python2.7/socket.py:571: error
#
</pre>

## Comments

Add a comment with a link to the failed test in the hope it will contain more useful information regarding the source of the error.

--- a/securedrop/tests/functional/functional_test.py
+++ b/securedrop/tests/functional/functional_test.py
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 
 from datetime import datetime
+import errno
 import mock
 from multiprocessing import Process
 import os
@@ -50,8 +51,26 @@ class FunctionalTest():
         return port
 
     def _create_webdriver(self, firefox, profile=None):
-        return webdriver.Firefox(firefox_binary=firefox,
-                                 firefox_profile=profile)
+        # see https://review.openstack.org/#/c/375258/ and the
+        # associated issues for background on why this is necessary
+        connrefused_retry_count = 3
+        connrefused_retry_interval = 5
+
+        for i in range(connrefused_retry_count + 1):
+            try:
+                driver = webdriver.Firefox(firefox_binary=firefox,
+                                           firefox_profile=profile)
+                if i > 0:
+                    # i==0 is normal behavior without connection refused.
+                    print('NOTE: Retried {} time(s) due to '
+                          'connection refused.'.format(i))
+                return driver
+            except socket.error as socket_error:
+                if (socket_error.errno == errno.ECONNREFUSED
+                        and i < connrefused_retry_count):
+                    time.sleep(connrefused_retry_interval)
+                    continue
+                raise
 
     def _prepare_webdriver(self):
         log_file = open(join(LOG_DIR, 'firefox.log'), 'a')

--- a/securedrop/tests/functional/functional_test.py
+++ b/securedrop/tests/functional/functional_test.py
@@ -49,14 +49,17 @@ class FunctionalTest():
         s.close()
         return port
 
-    def _create_webdriver(self):
+    def _create_webdriver(self, firefox, profile=None):
+        return webdriver.Firefox(firefox_binary=firefox,
+                                 firefox_profile=profile)
+
+    def _prepare_webdriver(self):
         log_file = open(join(LOG_DIR, 'firefox.log'), 'a')
         log_file.write(
             '\n\n[%s] Running Functional Tests\n' % str(
                 datetime.now()))
         log_file.flush()
-        firefox = firefox_binary.FirefoxBinary(log_file=log_file)
-        return webdriver.Firefox(firefox_binary=firefox)
+        return firefox_binary.FirefoxBinary(log_file=log_file)
 
     def setup(self):
         # Patch the two-factor verification to avoid intermittent errors
@@ -114,7 +117,7 @@ class FunctionalTest():
                 break
 
         if not hasattr(self, 'override_driver'):
-            self.driver = self._create_webdriver()
+            self.driver = self._create_webdriver(self._prepare_webdriver())
 
             # Poll the DOM briefly to wait for elements. It appears
             # .click() does not always do a good job waiting for the

--- a/securedrop/tests/pages-layout/functional_test.py
+++ b/securedrop/tests/pages-layout/functional_test.py
@@ -47,17 +47,11 @@ class FunctionalTest(functional_test.FunctionalTest):
             os.path.join(dirname(realpath(__file__)),
                          'screenshots', self.accept_languages))
         os.system("mkdir -p " + self.log_dir)
-        log_file = open(os.path.join(self.log_dir, 'firefox.log'), 'a')
-        log_file.write(
-            '\n\n[%s] Running Functional Tests\n' % str(
-                datetime.now()))
-        log_file.flush()
-        firefox = firefox_binary.FirefoxBinary(log_file=log_file)
+        firefox = self._prepare_webdriver()
         profile = webdriver.FirefoxProfile()
         profile.set_preference("intl.accept_languages", self.accept_languages)
         self.override_driver = True
-        self.driver = webdriver.Firefox(firefox_binary=firefox,
-                                        firefox_profile=profile)
+        self.driver = self._create_webdriver(firefox, profile)
         self._javascript_toggle()
 
         yield None

